### PR TITLE
[plugin-browserstack] Capture network logs

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: vividus-plugin-browserstack/src/main/resources/spring.xml [hunk 1]
```diff
@@ -1,11 +1,26 @@
...
 
+    <beans profile="mobile_app">
+        <bean class="org.vividus.browserstack.BrowserStackAppAutomateClient">
+            <constructor-arg index="0" value="${selenium.grid.username}" />
+            <constructor-arg index="1" value="${selenium.grid.password}" />
+        </bean>
+
+        <bean id="debugSteps" class="org.vividus.bdd.browserstack.steps.DebugSteps" />
+
+        <util:list id="stepBeanNames-BrowserStack" value-type="java.lang.String">
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/main/resources/spring.xml:22]: `stepBeanNames-BrowserStackMobileApp`

File: vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import com.browserstack.client.exception.BrowserStackException;
+
+import org.apache.commons.lang3.StringUtils;
+import org.vividus.util.UriUtils;
+import org.vividus.util.json.JsonUtils;
+import org.vividus.util.wait.DurationBasedWaiter;
+import org.vividus.util.wait.WaitMode;
+import org.vividus.util.wait.Waiter;
+
+public class BrowserStackAppAutomateClient extends AppAutomateClient
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java:34]: do you need inheritance? can't you use composition here?

File: vividus-plugin-browserstack/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 3]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (valfirst) [inline vividus-plugin-browserstack/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: why is it needed?

File: vividus-util/build.gradle [hunk 4]
```diff
@@ -7,9 +7,9 @@ dependencies {
     api(group: 'com.fasterxml.jackson.core', name: 'jackson-core')
     api(group: 'com.fasterxml.jackson.core', name: 'jackson-databind')
     api(group: 'com.fasterxml.jackson.datatype', name: 'jackson-datatype-jdk8')
+    api(group: 'org.apache.commons', name: 'commons-lang3', version: versions.commonsLang3)
```
Comment (valfirst) [inline vividus-util/build.gradle:10]: this change should go as a separate commit

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 5]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
 * See xref:tests-development:tests-configuration.adoc#_selenium_grid[Selenium Grid] properties to configure generic options.
 * Browser Stack specific properties must be prepended with `selenium.grid.capabilities.bstack\:options.` prefix, meaning given a property `projectName` with a value `My Project` the resulting property will be `selenium.grid.capabilities.bstack\:options.projectName=My Project`
 * Appium specific properties in Browser Stack must be prepended with `selenium.grid.capabilities.appium\:options.` prefix, meaning given a property `clearSystemFiles` with a value `true` the resulting property will be `selenium.grid.capabilities.appium\:options.clearSystemFiles=true`
+
+== Mobile Steps
+
+=== Save network logs
+
+==== *_Info_*
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:54]: please follow structure similar to this one: https://docs.vividus.dev/vividus/0.2.11-SNAPSHOT/plugins/plugin-rest-api.html#_init_multipart_http_request

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 6]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+
+Saves network logs from an application sesssion.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:56]: not clear

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 7]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+The application session must be closed before network logs can be saved.
+
+The following options available to enable network logs capturing on BrowserStack side:
+
+* Add `selenium.grid.capabilities.bstack:options.networkLogs=true` property into tests configuration, this will apply network logs capturing for all stories in suite.
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:61]: `apply ` -> `enable`

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 8]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+* Add `bstack:options.networkLogs` property with `true` value as parameter into start application step
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:62]:
```
Pass the capability with name `bstack:options.networkLogs` and value `true` as parameter to the step starting native mobile application.
``

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 9]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+.Enable network logs in start application step
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:63]: .Enable network logs in the step starting application

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 10]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+[source,gherkin]
+----
+Given I start mobile application with capabilities:
+|name                      |value|
+|bstack:options.networkLogs|true |
+----
+
+==== *_Wording_*
+
+[source,gherkin]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:73]: it should go after the first sentence: `Saves ...`

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 11]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+----
+When I save BrowserStack network logs to $scopes variable `$variableName`
+----
+
+==== *_Parameters_*
+
+. `$scopes` - xref:parameters:variable-scope.adoc[Variable Scopes]
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:80]: `Variable Scopes` -> `variable scopes`

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 12]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+. `$variableName` - name to prefix text content references
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:81]: not clear

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 13]
```diff
@@ -46,3 +46,45 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+
+==== *_Usage_*
+
+.Save network logs into variable
+[source,gherkin]
+----
+When I close mobile application
+When I save BrowserStack network logs to SCENARIO variable `networkLogs`
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:89]: any example on how to use it?

**Review by valfirst — CHANGES_REQUESTED** [review]

File: vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+{
+    private static final int DEFAULT_RETRY = 5;
+    private static final int DEFAULT_SECONDS = 25;
+
+    private final JsonUtils jsonUtils = new JsonUtils();
```
[the remark this answers is not in the document]
Comment (uarlouski) [reply vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java:39]: we need this just to handle simple corner case, I think it's better to pass into the client only those values that are logically suitable for it, like username and password, in case of jsonutil it's not clear why we even need it

File: vividus-plugin-browserstack/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker (hunk 3 above)
Comment (uarlouski) [reply vividus-plugin-browserstack/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: https://github.com/mockito/mockito/wiki/What%27s-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods

File: vividus-util/build.gradle (hunk 4 above)
Comment (uarlouski) [reply vividus-util/build.gradle:10]: this is related to my changes, otherwise we get compilation error

File: vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (valfirst) [reply vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java:39]: instantiation of ObjectMapper is quite expensive, that's why it's better to reuse the bean

File: vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java (hunk 2 above)
Comment (uarlouski) [reply vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java:34]: This client will be used later for getting logs for running test suite

File: vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (uarlouski) [reply vividus-plugin-browserstack/src/main/java/org/vividus/browserstack/BrowserStackAppAutomateClient.java:39]: so between ugly api and 0.001 sec saving we use the second choice, right?

File: docs/modules/plugins/pages/plugin-browser-stack.adoc [hunk 15]
```diff
@@ -46,3 +46,38 @@ Explore https://www.browserstack.com/app-automate/capabilities?tag=w3c[Capabilit
...
+When I save BrowserStack network logs to $scopes variable `$variableName`
+----
+
+. `$scopes` - xref:parameters:variable-scope.adoc[variable scopes]
+. `$variableName` - name of variable to save network logs
+
+The following options available to enable network logs capturing on BrowserStack side:
+
+* Add `selenium.grid.capabilities.bstack:options.networkLogs=true` property into tests configuration, this will enable network logs capturing for all stories in suite.
+* Add `@capability.bstack:options.networkLogs true` meta for a story being ran
```
Comment (valfirst) [inline docs/modules/plugins/pages/plugin-browser-stack.adoc:67]: `ran` -> `run`

**Review by valfirst — CHANGES_REQUESTED** [review]
And what about system tests?

File: vividus-tests/src/main/resources/properties/suite/system/mobile_app/ios/suite.properties [hunk 16]
```diff
@@ -1 +1 @@
-bdd.batch-1.meta-filters=groovy: !targetPlatform || targetPlatform.contains('ios')
+bdd.batch-1.meta-filters=groovy: (!targetPlatform || targetPlatform.contains('ios')) && !healthCheck
```
Comment (valfirst) [inline vividus-tests/src/main/resources/properties/suite/system/mobile_app/ios/suite.properties:1]: I believe it's better to create a new story, like: HealthCheckMobileAppSteps.story

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

**Review by ikalinin1 — APPROVED** [review]
